Add compile depth benchmarking for all models, run weekly.

jameszianxuTT opened the pull request:
### Ticket
#443 

### Problem description
Current tt-torch CI is meant to catch regressions, and our test coverage so far shows only what we expect to work. There is no infra to benchmark how far a model gets in actuality.

### What's changed
- Created an optional pytest CLI option, `--crashsafe`, to overload the junitxml record_property with an incremental XML logger that will transactionally create and flush an JUnitXML test report regardless of how the pytest crashes. It is resistant to segfaults, exceptions, assertions and other unexpected process terminations, and will create a JUnitXML test report parsable by the existing pytest parser in almost all cases (except if the test hangs CI, is skipped, or fails the ModelTester constructor).
- Created utilities to reconstruct the crashsafe JUnitXML test report into a parsable format, by injecting missing fields (eg. tags) that could not be flushed during a crash, and repair other possible report malformation. These utilities also merge the crashsafe JUnitXML reports into a single one, to appear transparent to the existing test data collection workflow
- Created a workflow to run the benchmark, 2H after nightlies start, on a Sunday, and generate an XLSX report 
- Added hooks in existing backend to record when a model achieves a given compile depth (SHLO, TTNN_IR, EXECUTE, PASS)

### Checklist
- [ ] New/Existing tests provide coverage for changes

jameszianxuTT commented:
Verification on correctness:
- A [reference run](https://github.com/tenstorrent/tt-torch/actions/runs/14020910456) on all models except musicgen (which hangs CI in full-eval).
  - In this reference run, all models that we currently test in EXECUTE through Nightly workflows are recorded, in the benchmark, as making it to EXECUTE. 
  - Similarly, all models that we currently test in COMPILE_TTNN_IR through Nightly workflows make it at least to EXECUTE in the benchmark.
- A [report generation mini run](https://github.com/tenstorrent/tt-torch/actions/runs/14061398658/job/39373541350) generated an XLSX summary, indicating that the following tests are promotable as they make it to EXECUTE in the benchmark (and this can be verified in the logs), while our existing test coverage does not test them in EXECUTE
  - `tests/models/torchvision/test_torchvision_object_detection.py::test_torchvision_object_detection[full-ssdlite320_mobilenet_v3_large-eval]` (tested in compile only)
  - `tests/models/yolov5/test_yolov5.py::test_yolov5[full-eval]` (not tested in CI)

jameszianxuTT commented on tests/conftest.py:
```diff
@@ -79,3 +93,94 @@ def record_test_timestamp(record_property):
     record_property("end_timestamp", end_timestamp)
+
+
+import xml.etree.ElementTree as ET
+import socket
```
I'll move these imports up

jameszianxuTT commented on .github/workflows/depth-benchmarks.yml:
```diff
@@ -0,0 +1,36 @@
+
+on:
+  schedule:
+    - cron: '0 6 * * 0'  # Runs at 06:00 UTC every Sunday
```
Yes, nightly tests run from midnight ET for around 4-6 hours, but I'm not sure what proportion of that time all our runners are saturated. 

I could try to run depth benchmarks starting from 6AM ET (10 UTC) but it risks starving out devs on CI for a few hours on Sunday morning, until we get some better priority scheduling.

jameszianxuTT replied:
Moved for now to 10 AM UTC.

kmabeeTT commented:
Any chance there is a reporting problem here, since this model is already in CI (added last night at 44203de, included in most recent nightly):

```
tests/models/yolov5/test_yolov5.py::test_yolov5[full-eval] (not tested in CI)
```

jameszianxuTT commented:
> Any chance there is a reporting problem here, since this model is already in CI (added last night at https://github.com/tenstorrent/tt-torch/commit/44203de347cbe472ea72bcf854d6fd97efde78dc, included in most recent nightly)

Not a reporting issue - This is a bit of a race condition with when I rebased / when you pushed your change / when I ran this experiment. My logs show 13 models in the full execute nightly tests, which is the correct number before your change. I probably didn't rebase in your changes to my branch at the time I ran my CI experiment, so it was using the old version of the file.

kmabeeTT commented:
> Not a reporting issue - This is a bit of a race condition with when I rebased / when you pushed your change / when I ran this experiment. 

Ah, got it! Scrolled back through your history run and saw you were based off main 2bb63d4 - though hmm, yolov5 was in compile e2e list at that time, and I'd think more models should have been flagged as available to be promoted to execute since there were indeed more than 7 or so at that time (I added 7 in my commit a bit later).

jameszianxuTT commented:
> Ah, got it! Scrolled back through your history run and saw you were based off main [2bb63d4](https://github.com/tenstorrent/tt-torch/commit/2bb63d47d91753e19bd73d27f9d4ba3a36c30abe) - though hmm, yolov5 was in compile e2e list at that time, 
<img width="935" alt="image" src="https://github.com/user-attachments/assets/50ff81a2-77d7-4c28-8023-3831838ac09c" />

My bad, I reversed those rows in my head and typed it in my comment incorrectly. The spreadsheet says yolov5 was in the compile e2e list, not mobilenet_V3_large. 

> and I'd think more models should have been flagged as available to be promoted to execute since there were indeed more than 7 or so at that time (I added 7 in my commit a bit later).

This is true, I only ran a small subset of the models in this experiment (35 benchmarked/140 total) to not monopolize the runners.

kmabeeTT commented:
> This is true, I only ran a small subset of the models in this experiment (35 benchmarked/140 total) to not monopolize the runners.

Cool, all makes sense now, thx!

jameszianxuTT commented:
Just added a testcase "tests/torch/tools/test_benchmark.py" to be run in the run-tests.yml on Push/on PR that extends those test times by 30 seconds. It's not the rate determining step though, so it shouldn't increase the overall onPush/onPR times.

It demonstrates some basic usage of the scripts I've added, and may be a useful reference.

kmabeeTT reviewed (approved):
In general, looks awesome! Let's see how it goes this weekend! 

Is the idea to add test_benchmark.py into push/pr CI flows later? Thx for adding changes, and that new ~test.

In the future maybe add some coloring or easier to spot difference between promotable yes vs no in the generated xls since from quick eyeball it's hard to spot the Yes's.

jameszianxuTT commented:
> Is the idea to add test_benchmark.py into push/pr CI flows later? 

This test is already run as part of the pytorch unit tests, including in the [check executed on this PR.](https://github.com/tenstorrent/tt-torch/actions/runs/14093856323/job/39477308582#step:11:674) 

Unfortunately you can't see the output without `pytest -s`, but it's pretty interesting as it runs MNIST in a subprocess and sigkills it mid-execution before reconstructing the partial log and checking that it's well formed.

You probably know this but the push/pr checks run `pytest -v tests/torch` and that test discovery is a recursive search - I've put my tests for profiler and benchmark into `tests/torch/tools` so they're autodetected. 

I will be moving those tests via #444 later.

> In the future maybe add some coloring 

Sure, I'll need to do a followup PR to collect the new tests that Het is adding this week and throw them into the test matrix, I can add the highlighting in with that PR
